src/wobjectcpp.h

```diff
@@ -0,0 +1,124 @@
+}
+
+#if __cplusplus <= 201700L
+template<bool... Bs>
+constexpr bool all() {
+    bool b = true;
+    w_internal::ordered2<bool>({(b = b && Bs)...});
+    return b;
+}
+#endif
+
+template<class F, int Flags, class IC = int, class ParamTypes = w_internal::StringViewArray<>, class ParamNames = w_internal::StringViewArray<>>
+struct MetaMethodInfoBuilder {
+    StringView name;
+    F func;
+    ParamTypes paramTypes{};
+    ParamNames paramNames{};
+    static constexpr int flags = Flags;
+    using IntegralConstant = IC;
+
+#if __cplusplus > 201700L
+    template<class... Args, class = std::enable_if_t<(std::is_same_v<std::decay_t<Args>, StringView> && ...)>>
+#else
+    template<class... Args, class = std::enable_if_t<all<std::is_same<std::decay_t<Args>, StringView>::value...>()>>
+#endif
+    constexpr auto setParamTypes(Args... paramTypes) const
+        -> MetaMethodInfoBuilder<F, Flags, IC, w_internal::StringViewArray<sizeof... (Args)>, ParamNames> {
+        return {name, func, {paramTypes...}, paramNames};
+    }
+#if __cplusplus > 201700L
+    template<class... Args, class = std::enable_if_t<(std::is_same_v<std::decay_t<Args>, StringView> && ...)>>
+#else
+    template<class... Args, class = std::enable_if_t<all<std::is_same<std::decay_t<Args>, StringView>::value...>()>>
+#endif
+    constexpr auto setParamNames(Args... paramNames) const
+        -> MetaMethodInfoBuilder<F, Flags, IC, ParamTypes, w_internal::StringViewArray<sizeof... (Args)>> {
+        return {name, func, paramTypes, {paramNames...}};
+    }
+    template<int... fs>
+    constexpr auto addFlags(w_internal::W_MethodFlags<fs>...) const
+#if __cplusplus > 201700L
+        -> MetaMethodInfoBuilder<F, (fs | ... | Flags), IC, ParamTypes, ParamNames> {
+#else
+        -> MetaMethodInfoBuilder<F, w_internal::summed<fs ..., Flags>, IC, ParamTypes, ParamNames> {
+#endif
+        return {name, func, paramTypes, paramNames};
+    }
+    template<class IC2>
+    constexpr auto setIntegralConstant() const -> MetaMethodInfoBuilder<F, Flags, IC2, ParamTypes, ParamNames> {
+        return {name, func, paramTypes, paramNames};
+    }
+
+    constexpr auto build() const -> w_internal::MetaMethodInfo<F, Flags, IC, ParamTypes, ParamNames> {
+        return {func, name, paramTypes, paramNames};
+    }
+};
+
+template<typename F>
+constexpr auto makeSignalBuilder(StringView name, F func) {
+    return MetaMethodInfoBuilder<F, W_MethodType::Signal.value>{name, func};
+}
+
+
+} // namespace w_cpp
+
+/// \macro W_CPP_PROPERTY(callback)
+/// allows to create multiple properties from a templated structure using regular C++.
+///
+/// example usage:
+///     template<size_t I>
+///     struct MyProperties;
+///     template<>
+///     struct MyProperties<0> {
+///         constexpr static auto property = w_cpp::makeProperty<QString>(w_cpp::viewLiteral("name"), w_cpp::viewLiteral("QString"))
+///             .setGetter(&Example::getName);
+///     };
+///     W_CPP_PROPERTY(MyProperties)

```

Comment:
Would it be possible to have amacro such as this would be possible:

```
constexpr static auto propertyName = w_cpp::makeProperty<QString>(w_cpp::viewLiteral("name"), w_cpp::viewLiteral("QString"))
￼	///             .setGetter(&Example::getName);

W_ADD_PROPERTY(propertyName)
```

I don't really understand the meaning in CPP (because everything is CPP, so I suggest ADD instead.
And I think hafing to specialize a template is a bit complicated, if there are several properties, one can call W_ADD_PROPERTY several time


Perhaps even we should allow:

```
W_ADD_PROPERTY(w_cpp::makeProperty<QString>(w_cpp::viewLiteral("name"), w_cpp::viewLiteral("QString")))
```
￼	
For adding several property at once, a tuple could be used.

Reply:
`W_CPP_` is just the prefix I used. It's basically the C++ way to add properties.

Specializing a template makes this approach very accessible. Users of this API most likely will have their own constexpr based way to represent the properties. We don't know what they use. So they have to provide an adapter.

Internally we count everything. So this approach is very light weight and adds almost no compile time.

If we want to allow a tuple based approach, we would force the compiler to create a complex constexpr enable tuple, just to decompose it into something that is indexed. This would make the API more complicated, harder to use and slower to compile.